ddd/ddd-05-mem-nonblocking.DISCUSSION.md

```diff
@@ -0,0 +1,250 @@
+    can do is to ensure that we accommodate this use case for applications, so
+    that there is some rougly analagous solution available to them. This may not
+    be "minimal changes" in the sense of "the changes are small in an absolute
+    sense" but it does meet the definition of "minimal changes" where "minimal"
+    means "as small as actually possible". While adaptation of this kind of
+    application (demo 5) will require more work than the other kinds of
+    application (as demonstrated in demos 1 through 4) we can at least provide
+    an interface which applications will find familiar and which lacks any
+    excessive impedence discontinuities with their own code.
+
+  - There already exists BIO_dgram, which is used with DTLS. This causes
+    BIO_read/BIO_write to work similarly to recv/send. This works on
+    a file descriptor; there is no analagous BIO_dgram_mem.
+    Such a BIO could be created, which would model datagram semantics
+    for read/write operations (i.e., a sequence of BIO_write calls
+    with given lengths results in an exactly matching sequence of
+    BIO_read calls with the same lengths returned).
+
+  - There are some caveats for this idea. In order for this to work:
+
+    - applications must ensure there is a guaranteed 1:1 correspondance
+      between send/recv and BIO_read/BIO_write calls; and
+
+    - the buffer passed to BIO_read() and recv() must always be big
+      enough to receive the largest possible packet, else silent
+      truncation occurs.
+
+    The latter is an unfortunate design issue with the Berkeley sockets API
+    for SOCK_DGRAM, which is replicated here in BIO_read which simply serves
+    as a wrapper for it. The caller thus has to know the worst-case UDP
+    datagram payload size, but knowing this requires knowledge of the
+    network or at least the configuration of local network interfaces.
+
+    (Random factoid: this design issue once in the past led me to determine
+    the maximum size of an incoming UDP datagram as
+
+      max(64 KiB, max(MTU of every local interface))
+
+    -- reason being that an IPv6 fragmentation header has a 16-bit length field,
+    so an IP packet greater than 64 KiB (using the IPv6 jumbogram header) can
+    only be sent if you have an interface with a native MTU exceeding it.
+    Another option is to use MSG_PEEK | MSG_TRUNC on systems which support it
+    but this requires calling recv twice per packet.)
+
+  - For the buffer size issue, our options for supporting existing applications
+    include:
+
+    - Using a maximum MTU of 1472 by default; then any application using
+      old BIO_read/BIO_write calls will work so long as it uses a buffer
+      of at least 1472 bytes (and there is a 1:1 correspondance of BIO calls and
+      syscalls).
+
+    - Applications which can support larger MTUs (e.g. a network with
+      jumbo frames) could indicate support via some method TBD, such as
+      capability flags (see below).
```

Comment:
Can't `BIO_dgram_mem` just support `BIO_CTRL_DGRAM_SET_MTU` in the same way that `BIO_dgram` does. The application can then set it to whatever value is appropriate (or otherwise let it default to 1472). `BIO_dgram_mem` should support the same MTU querying ctrls that `BIO_dgram` supports, e.g. `BIO_CTRL_DGRAM_QUERY_MTU`